usecase/usecase_test.go

```diff
@@ -0,0 +1,144 @@
+package usecase_test
+
+import (
+	"errors"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"peanut/domain"
+)
+
+var _ = Describe("Usecase", func() {
+
+	var book domain.Book
+	var updatingBookForm domain.UpdateBookForm
+	var bookID int
+	BeforeEach(func() {
+		book = domain.Book{
+			Name:  "test",
+			Year:  1999,
+			Price: 19.99,
+		}
+		//updateBookForm
+		updatingBookForm.Name = &book.Name
+		updatingBookForm.Year = &book.Year
+		updatingBookForm.Price = nil
+		bookID = 1
+	})
+
+	Describe("API get book by id", func() {
+		Context("Found book", func() {
+			It("Should be success", func() {
+				//prepare
+				bookRepo.EXPECT().GetBookById(bookID).Return(&book, nil)
+				//do
+				b, err := bookUc.GetBook(bookID)
+				//check
+				Expect(err).To(BeNil())
+				Expect(b).NotTo(BeNil())
+			})
+		})
+		Context("Not found book", func() {
+			It("Should be failed", func() {
+				//prepare
+				bookRepo.EXPECT().GetBookById(bookID).Return(nil, errors.New("Book not found"))
+				//do
+				b, err := bookUc.GetBook(bookID)
+				//check
+				Expect(err).Should(HaveOccurred())
+				Expect(b).To(BeNil())
+			})
+		})
+	})
+
+	Describe("API create book", func() {
+		Context("Create book success", func() {
+			It("Should be success", func() {
+				//prepare
+				bookRepo.EXPECT().CreateBook(book).Return(&book, nil)
+				//do
+				_, err := bookUc.CreateBook(book)
+				//check
+				Expect(err).To(BeNil())
+			})
+		})
+		Context("Create book failed", func() {
+			It("Should be failed", func() {
+				//prepare
+				bookRepo.EXPECT().CreateBook(book).Return(nil, errors.New("username exist"))
+				//do
+				_, err := bookUc.CreateBook(book)
+				//check
+				Expect(err).Should(HaveOccurred())
+			})
+		})
+	})
+
+	Describe("Api update book", func() {
+		Context("Found book, update book success", func() {
```

Comment:
chỗ này em có thể gộp vào 1 context
```
Context("Found book", func() {
  It("should be updated successfully", func(){})
  It("should be updated failed", func(){})
})
```

Reply:
ở trong context("found book") thì em cho fake 
`bookRepo.EXPECT().GetBookById(bookID).Return(&book, nil)`
còn trong 2 cái It thì em 
chia 2 trường hợp
`bookRepo.EXPECT().UpdateBook(book, bookID).Return(&book, nil)`
`bookRepo.EXPECT().UpdateBook(book, bookID).Return(nil, errors.New("Internal error"))`
=> e thử xem cách này đc k? a cũng chưa thử

Reply:
để ở context không test được ạ, phải để ở cả 2 cái It test mới chạy được